src/main/java/VectorClockBackend.java

```diff
@@ -0,0 +1,142 @@
+    public OtpErlangBinary update(OtpErlangBinary JavaObjectId, OtpErlangBinary binary) throws NoSuchObjectException {
+        if (!ObjectTable.containsKey(JavaObjectId)) {
+            try {
+                CRDT crdt_object = ((CRDTEffect) binary.getObject()).crdt;
+                Snapshot mapentry = new Snapshot(crdt_object, new TreeSet<GenericEffect>());
+                ObjectTable.put(JavaObjectId, mapentry);
+            } catch (ClassCastException e) {
+                // We don't have the object and we have been given an update
+                // so we need to request the object from antidote and try again
+                assert (binary.getObject().getClass() == GenericFunction.class);
+                throw new NoSuchObjectException();
+            }
+        } else {
+            try {
+                GenericEffect updateEffect = (GenericEffect) binary.getObject();
+                TreeSet<GenericEffect> sortedEffectSet = ObjectTable.get(JavaObjectId).effectbuffer;
+                sortedEffectSet.add(updateEffect);
+            } catch (ClassCastException e) {
+                /* intentionally ignore this exception */
+                // This happens because we don't have an Idset for crdt initializations through
+                // updates so we got a redundant crdt object(something we already have)
+                assert (binary.getObject().getClass() == CRDT.class);
+            }
+        }
+
+        return JavaObjectId;
+    }
+
+    @Override
+    public OtpErlangBinary downstream(OtpErlangBinary JavaObjectId, OtpErlangBinary binary, OtpErlangMap clock,
+            OtpErlangMap global_clock) {
+        VectorClock effectClock = new VectorClock(clock);
+        GlobalClockTime = new VectorClock(global_clock);
+
+        OtpErlangBinary bin;
+        try {
+            CRDTEffect effect = new CRDTEffect((CRDT) binary.getObject(), effectClock);
+            bin = new OtpErlangBinary(effect);
+        } catch (ClassCastException e) {
+            GenericEffect effect = new GenericEffect((GenericFunction) binary.getObject(), effectClock);
+            bin = new OtpErlangBinary(effect);
+        }
+
+        return bin;
+    }
+
+    @Override
+    public OtpErlangTuple snapshot(OtpErlangBinary JavaObjectId) throws NoSuchObjectException {
+        // assert that object is in table else request it
+        Snapshot mapentry = ObjectTable.get(JavaObjectId);
+        CRDT crdt_object = mapentry.object;
+        if (crdt_object == null) {
+            throw new NoSuchObjectException();
+        }
+
+        // Apply any effects that have passed the clock time to the new copy of the
+        // object
+        TreeSet<GenericEffect> crdt_effect_buffer = mapentry.effectbuffer;
+        TreeSet<GenericEffect> new_crdt_effect_buffer = new TreeSet<GenericEffect>();
+        for (GenericEffect e : crdt_effect_buffer) {
+            // Use a correct compare based on types
+            if (e.time.lessthan(this.GlobalClockTime)) {
```

Comment:
e.time.lessthan(this.GlobalClockTime) || (0 == e.time.compareTo(this.GlobalClockTime))

Reply:
Sounds good